sigstore-java/src/main/java/dev/sigstore/StringMatchable.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2024 The Sigstore Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.sigstore;
+
+import com.google.common.base.Preconditions;
+import java.util.regex.Pattern;
+
+/**
+ * An interface for allowing direct string matching or regular expressions. Use the static factory
+ * {@link #of(String)} or {@link #of(Pattern)} to instantiate the appropriate matcher.
+ */
+public interface StringMatchable {
```

Comment:
The naming sounds a bit odd. Guava uses `CharMatcher implements Predicate<Character>`

Reply:
I mean this matches more the interface naming (vs class naming) `Runnable`, `Callable`, etc. Vs `CharMatcher` which is an abstract class? I'm not super tied to this name, I just want this functionality.

Reply:
Compare with `java.util.function.Predicate` (vs `Predicatable` or `Testable`).

Frankly, this sounds like `StringPredicate extends Predicate<String>`.
At the same time, if protobuf-specs include a data-class-like (`regexp` and `literal` values), then we should probably expose it in the same way here.
Of course, `sealed interface` is not yet available, however we can expose it with protobuf-generated classes or just two case classes like `LiteralId` and `RegexId`.

Reply:
I don't think protobuf generated classes will be in the API. So we don't want to do that, I can rename a lot of these maybe. Lemme play around with it.